zerver/lib/users.py

```diff
@@ -524,6 +653,76 @@ def get_cross_realm_dicts() -> List[Dict[str, Any]]:
     return result
 
 
+def get_data_for_inaccessible_user(
+    user_id: int, user_is_active: bool, realm: Realm
+) -> Dict[str, Any]:
+    user_email = Address(username=f"user{user_id}", domain=get_fake_email_domain(realm)).addr_spec
+
+    # We just set date_joined field to UNIX epoch.
+    user_date_joined = timestamp_to_datetime(0)
+
+    user_dict = dict(
+        email=user_email,
+        user_id=user_id,
+        avatar_version=1,
+        is_admin=False,
+        is_owner=False,
+        is_guest=False,
+        is_billing_admin=False,
+        role=UserProfile.ROLE_MEMBER,
+        is_bot=False,
+        full_name="Unknown user",
+        timezone="",
+        is_active=user_is_active,
+        date_joined=user_date_joined.isoformat(),
+        delivery_email=None,
+        avatar_url=get_avatar_for_inaccessible_user(),
+        profile_data={},
+    )
+    return user_dict
+
+
+def get_accessible_user_ids(realm: Realm, user_profile: UserProfile) -> List[int]:
+    subscribers_dict_of_target_user_subscriptions = get_subscribers_of_target_user_subscriptions(
+        [user_profile]
+    )
+    users_involved_in_dms_dict = get_users_involved_in_dms_with_target_users([user_profile], realm)
+
+    accessible_user_ids = (
+        {user_profile.id}
+        | subscribers_dict_of_target_user_subscriptions[user_profile.id]
+        | users_involved_in_dms_dict[user_profile.id]
+    )
```

Comment:
We should probably add a comment noting that this doesn't include cross-realm bots and why that's correct (presumably, because they're sent elsewhere in the API).

Reply:
Added a comment here about bots, but we would need to include bots here as well by using `include_bots` parameter.